keps/sig-scheduling/20190317-non-preempting-priorityclass

```diff
@@ -0,0 +1,104 @@
+---
+title: Add NonPreempting Option For PriorityClasses
+authors:
+  - "@vllry"
+owning-sig: sig-scheduling
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "k82cn"
+  - "wgliang"
+approvers:
+  - "bsalamat"
+editor: Vallery Lancey
+creation-date: 2019-03-17
+last-updated: 2019-03-17
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Allow PriorityClasses To Be Non-Preempting
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+[PriorityClasses](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/) are a beta feature,
+which impact the scheduling and eviction of pods.
+Pods will be scheduled according to descending priority.
+If a pod cannot be scheduled due to insufficient resources,
+lower-priority pods will be descheduled to make room.
+
+This proposal makes the pre-empting (descheduling) behavior optional,
+by adding a new field to PriorityClasses.
+If a PriorityClass does not have preemption enabled,
+the scheduler will not preempt pods in order to schedule a pod of that priority.
+
+## Motivation
+
+High-priority, non-preempting workloads are a common data science use case.
+Preempting batch workloads is a waste, as the work unit must be repeated.
+
+### Goals
+
+Add a boolean to PriorityClasses,
```

Comment:
Do we also need to add this field into Pod? For now, PriorityClass will be translate into `spec.priority` in pod for scheduling.

Reply:
BTW, please also highlight whether kubelet should consider this option/configuration :)

Reply:
I'll have to double check the code.

Reply:
numeric priority was denormalized from priority class into the pod at admission time, which:
* avoids issues with mutation of priorityclass objects
* ensures the scheduler only has to pay attention to pods
* allows kubelets to set numeric priority on static pods

is there a reason not to do the same here?

Reply:
IIUC, the idea that @k82cn and @liggitt are proposing is to add a "Preempting" field to PodSpec instead of PriorityClass. Then a user sets the field at the time of pod creation and its default value will be true. Is my understanding correct?
I like this idea, because as Jordan pointed out, the scheduler does not need to look at PriorityClasses. It also gives us the flexibility to use the same PriorityClass for two sets of pods where one set can preempt others and the other cannot.
The downside of this approach is that we lose the ability to grant preempting quota to users. ResourceQuota supports PriorityClass. An admin can give a certain amount of quota to a user at a specific priority identified by a PriorityClass. When the "Preempting" field is in PriorityClass, an admin can choose to give users preempting or non-preempting priority, but if the field in the PodSpec, the admin will not be able to choose between the two options anymore. The admin will only be able to assign quota at a specific priority and it will be up to users to decide whether they want to create pods as preempting or non-preempting.

An alternative approach is to add the field to both PriorityClass and PodSpec. Similar to the "Priority" field of PodSpec, the "preempting" field is populated from the PriorityClass by the "Priority" admission controller at the time of pod creation. This approach has the benefits that Jordan pointed out without the downside that I mentioned above.

Reply:
Thanks, that makes sense.
cc @denkensk

Reply:
So, are we going to have `Preempting` field in both podSpec and PriorityClass?